nucliadb_utils/nucliadb_utils/storages/pg.py

```diff
@@ -0,0 +1,560 @@
+
+# Table design notes
+# - No foreign key constraints ON PURPOSE
+# - No cascade handling ON PURPOSE
+CREATE_TABLE = """
+CREATE TABLE IF NOT EXISTS kb_files (
+    kb_id TEXT,
+    file_id TEXT,
+    filename TEXT,
+    size INTEGER,
+    content_type TEXT,
+    PRIMARY KEY(kb_id, file_id)
+);
+
+CREATE TABLE IF NOT EXISTS kb_files_fileparts (
+    kb_id TEXT,
+    file_id TEXT,
+    part_id INTEGER,
+    size INTEGER,
+    data BYTEA,
+    PRIMARY KEY(kb_id, file_id, part_id)
+);
+"""
+
+
+class FileInfo(TypedDict):
+    filename: str
+    size: int
+    content_type: str
+    key: str
+
+
+class ChunkInfo(TypedDict):
+    part_id: int
+    size: int
+
+
+class Chunk(ChunkInfo):
+    data: bytes
+
+
+class PostgresFileDataLayer:
+    """
+    Responsible for interating with the database and
+    abstracting any sql and connection management.
+    """
+
+    def __init__(self, connection: asyncpg.Connection):
+        self.connection = connection
+
+    async def initialize_kb(self, kbid: str) -> bool:
+        # there's really no record keeping or init
+        # per kb that we care to do
+        return True
+
+    async def delete_kb(self, kbid: str) -> bool:
+        async with self.connection.transaction():
+            await self.connection.execute(
+                """
+DELETE FROM kb_files
+WHERE kb_id = $1
+""",
+                kbid,
+            )
+            await self.connection.execute(
+                """
+DELETE FROM kb_files_fileparts
+WHERE kb_id = $1
+""",
+                kbid,
+            )
+        return True
+
+    async def create_file(
+        self, *, kb_id: str, file_id: str, filename: str, size: int, content_type: str
+    ) -> None:
+        async with self.connection.transaction():
+            await self.connection.execute(
+                """
+INSERT INTO kb_files (kb_id, file_id, filename, size, content_type)
+VALUES ($1, $2, $3, $4, $5)
+""",
+                kb_id,
+                file_id,
+                filename,
+                size,
+                content_type,
+            )
+
+    async def delete_file(self, kb_id: str, file_id: str) -> None:
+        async with self.connection.transaction():
+            await self.connection.execute(
+                """
+DELETE FROM kb_files
+WHERE kb_id = $1 AND file_id = $2
+""",
+                kb_id,
+                file_id,
+            )
+            await self.connection.execute(
+                """
+DELETE FROM kb_files_fileparts
+WHERE kb_id = $1 AND file_id = $2
+""",
+                kb_id,
+                file_id,
+            )
+
+    async def append_chunk(self, *, kb_id: str, file_id: str, data: bytes) -> None:
+        async with self.connection.transaction():
+            await self.connection.execute(
+                """
+INSERT INTO kb_files_fileparts (kb_id, file_id, part_id, data, size)
+VALUES ($1, $2, (SELECT COALESCE(MAX(part_id), 0) + 1 FROM kb_files_fileparts WHERE kb_id = $1 AND file_id = $2), $3, $4)
```

Comment:
note: will `part_id` start always at 1? I guess it doesn't really matter?

Reply:
yes, I can change it to work without or a coalesce or just with 1 separate query instead so it's more we don't start at 1